### PR TITLE
Fix BigInteger parsing of substring span

### PR DESCRIPTION
**Pull request by stephentoub**
Line-by-line port of https://github.com/dotnet/coreclr/pull/17808.

(This code is almost identical; it'd be nice to unify on the source shared from CoreLib.)

Fixes https://github.com/dotnet/corefx/issues/35176
cc: @tannergooding, @jkotas

### CODE REVIEW

**Comment by stephentoub** [comment]
> Could you log an issue tracking us merging this Number class with the main Number class?

Yup. https://github.com/dotnet/corefx/issues/35189

File: src/System.Runtime.Numerics/tests/BigInteger/parse.netcoreapp.cs [hunk 1]
```diff
@@ -9,6 +9,24 @@ namespace System.Numerics.Tests
 {
     public partial class parseTest
     {
+        [Theory]
+        [InlineData("123456789", 0, 9, "123456789")]
+        [InlineData("123456789", 0, 1, "1")]
+        [InlineData("123456789", 1, 3, "234")]
```
Comment (danmosemsft) [inline src/System.Runtime.Numerics/tests/BigInteger/parse.netcoreapp.cs]: `[InlineData("123456789", 8, 1, "9")]` maybe (non zero offset, ends at end)?

**Comment by danmosemsft** [comment]
Do we need to backport?

**Comment by stephentoub** [comment]
@danmosemsft: https://github.com/dotnet/corefx/pull/35194

**Pull request merged**